provisioner/ansible/provisioner.go

```diff
@@ -307,50 +300,117 @@ func (p *Provisioner) Provision(ctx context.Context, ui packer.Ui, comm packer.C
+	tf, err := ioutil.TempFile(p.config.InventoryDirectory, "packer-provisioner-ansible")
+	if err != nil {
+		return fmt.Errorf("Error preparing inventory file: %s", err)
+	}
 
-	if len(p.config.InventoryFile) == 0 {
-		tf, err := ioutil.TempFile(p.config.InventoryDirectory, "packer-provisioner-ansible")
+	host := fmt.Sprintf("%s ansible_host=%s ansible_user=%s ansible_port=%d\n",
+		hostIP, p.config.HostAlias, p.config.User, hostPort)
+	if p.ansibleMajVersion < 2 {
+		host = fmt.Sprintf("%s ansible_ssh_host=%s ansible_ssh_user=%s ansible_ssh_port=%d\n",
+			hostIP, p.config.HostAlias, p.config.User, hostPort)
+	}
+
+	w := bufio.NewWriter(tf)
+	w.WriteString(host)
+	for _, group := range p.config.Groups {
+		fmt.Fprintf(w, "[%s]\n%s", group, host)
+	}
+
+	for _, group := range p.config.EmptyGroups {
+		fmt.Fprintf(w, "[%s]\n", group)
+	}
+
+	if err := w.Flush(); err != nil {
+		tf.Close()
+		os.Remove(tf.Name())
+		return fmt.Errorf("Error preparing inventory file: %s", err)
+	}
+	tf.Close()
+	p.config.InventoryFile = tf.Name()
+
+	return nil
+}
+
+func (p *Provisioner) Provision(ctx context.Context, ui packer.Ui, comm packer.Communicator, generatedData map[string]interface{}) error {
+	ui.Say("Provisioning with Ansible...")
+	// Interpolate env vars to check for generated values like password and port
+	p.generatedData = generatedData
+	p.config.ctx.Data = generatedData
+	for i, envVar := range p.config.AnsibleEnvVars {
+		envVar, err := interpolate.Render(envVar, &p.config.ctx)
 		if err != nil {
-			return fmt.Errorf("Error preparing inventory file: %s", err)
+			return fmt.Errorf("Could not interpolate ansible env vars: %s", err)
 		}
-		defer os.Remove(tf.Name())
-
-		host := fmt.Sprintf("%s ansible_host=127.0.0.1 ansible_user=%s ansible_port=%d\n",
-			p.config.HostAlias, p.config.User, p.config.LocalPort)
-		if p.ansibleMajVersion < 2 {
-			host = fmt.Sprintf("%s ansible_ssh_host=127.0.0.1 ansible_ssh_user=%s ansible_ssh_port=%d\n",
-				p.config.HostAlias, p.config.User, p.config.LocalPort)
+		p.config.AnsibleEnvVars[i] = envVar
+	}
+	// Interpolate extra vars to check for generated values like password and port
+	for i, arg := range p.config.ExtraArguments {
+		arg, err := interpolate.Render(arg, &p.config.ctx)
+		if err != nil {
+			return fmt.Errorf("Could not interpolate ansible env vars: %s", err)
 		}
+		p.config.ExtraArguments[i] = arg
+	}
+
+	// Set up proxy if there's no host IP to access, regardless of user config.
+	hostIP := generatedData["Host"]
+	if hostIP == "" && p.config.UseProxy.False() {
+		ui.Error("Warning: use_proxy is false, but instance does" +
+			" not have an IP address to give to Ansible. Falling back" +
+			" to use localhost proxy.")
+		p.config.UseProxy = confighelper.TriTrue
+	}
 
-		w := bufio.NewWriter(tf)
-		w.WriteString(host)
-		for _, group := range p.config.Groups {
-			fmt.Fprintf(w, "[%s]\n%s", group, host)
+	privKeyFile := ""
+	if !p.config.UseProxy.False() {
```

Comment:
Oh I got confused for a bit but now I get it; we will always start it unless it's explicitly set to false. 👍

Reply:
Do you think it would be a good idea to switch the content of two blocks ?
```suggestion
	if p.config.UseProxy.False() == true {
	    ui.Message("Not using Proxy adapter for Ansible run...")
	} else {
```

Reply:
That might be more intuitive 🤔

Reply:
Thinking on it more, I don't think this is any better. For me the original way is more intuitive, so I think this is just gonna be a personal preference thing.

Reply:
Got it 👍 🙂 yeah that's personal !